bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -1049,6 +1049,42 @@ impl Connection {
     pub fn resumed(&self) -> bool {
         unsafe { s2n_connection_is_session_resumed(self.connection.as_ptr()) == 1 }
     }
+
+    /// Associates an arbitrary application context with the Connection to be later retrieved via
+    /// the [`Self::application_context()`] and [`Self::application_context_mut()`] APIs.
+    ///
+    /// This API will override an existing application context set on the Connection.
+    pub fn set_application_context<T: Send + Sync + 'static>(&mut self, app_context: T) {
+        self.context_mut().app_context = Some(Box::new(app_context));
+    }
+
+    /// Retrieves a reference to the application context associated with the Connection.
+    ///
+    /// To set a context on the connection, use the [`Self::set_application_context()`] API. If an
+    /// application context of the specified type hasn't already been set on the Connection, None
+    /// will be returned.
+    ///
+    /// To retrieve a mutable reference to the context, use [`Self::application_context_mut()`].
+    pub fn application_context<T: Send + Sync + 'static>(&self) -> Option<&T> {
+        match self.context().app_context.as_ref() {
+            None => None,
+            Some(app_context) => app_context.downcast_ref::<T>(),
+        }
```

Comment:
Should requesting the wrong type be None, or an error? I'm not sure I like "wrong type" and "no context" being indistinguishable.

Reply:
I changed this to return an error when the wrong type is requested, but it feels strange because the application always has to provide a type to the getter, so it feels like the wrong type is provided in both cases. But if the ability to distinguish between these cases seems helpful then it makes sense.

Reply:
>  it feels strange because the application always has to provide a type to the getter, so it feels like the wrong type is provided in both cases

My thinking is that there can only be one context. If you set a context of typeA and then try to get a context of typeB, that seems like an error. Like I'd expect that was a coding mistake instead of intentional. If you did the C equivalent, then we'd pass the void context pointer back cast to the wrong type and everything would break.

But it's fair to say this isn't C and it doesn't break everything, so maybe None is fine? We might want other opinions.

Reply:
I'd vote for committing to `Result` even harder 😄, and using `Result<&T>` instead of `Result<Option<&T>>`.